NOTES

afraser opened the pull request:
Just making this PR for the sake of having context for notes

keegsands committed:
Added summary and detail page

- First major work for the podcast sample app.
Small Fixes

Updated to use different feed and fixed resulting issues in the UI
Clean up warnings

Fixed some warnings to the way states were loaded.
Convert to XML

Changed the feed reader to look for XML instead of JSON because most podcasts don't support the feed.json format yet.
Flipped list order

- Episodes now display in reverse order
- Styling fixes for font and image borders

keegsands committed:
Fixed class attribute warning

Need to use className instead of class for JSX

keegsands committed:
Function Component Conversion

Converted the main components over to be Function Components.

keegsands committed:
Abstracted out Model

Abstracting out the podcast information into a generic object.
Broke out header component
Abstraction Continues

Moved parsing of podcast XML into a util class and also simplified passing properties around so a podcast object is passed instead of multiple text properties.
More Abstraction

Abstracted the episodes list of the podcast into the podcast-util and update Detail to use the new structure
Finishing Touches

Adding comments
Stopping potential reload issues
Moving reverse into parsing so the view doesn't have to worry about the order

keegsands committed:
Episode Order Changes

- No longer reversing order of episodes
- Update title of the page to match the podcast name
- New button for flipping the order of the episodes

keegsands committed:
Routing Support

Updating the names of components to prepare for new routing feature.
Refactoring for Routing

More moving and shaking of files  leading up to roting
Creating Screens

Created and started using Screens for displaying components
Implement Router

Added basic router functionality
Router Work

Center the contents of the list view
Cleaning up navigation


More visual clean up

Fixed artifact displaying on initial load of podcast page
Fixed link size and color
Routing Support

Adding routing support for displaying multiple podcasts.  Using convention over configuration to start.
Routing Support

Removed unused Redirect

keegsands committed:
Fix Error with file name case

keegsands committed:
Post remote repo work

Updating the readme

keegsands committed:
Post remote repo work

Removing some feeds and replacing them with bogus feeds.

keegsands committed:
Global state

Prepping for global state

afraser commented on src/components/podcast/Podcast.js:
```diff
@@ -0,0 +1,63 @@
+    // Track if the podcast has been loaded
+    const [loaded, setLoaded] = useState(false);
+
+    useEffect(() => { loadFeed(props.feed + '.xml') });
```
Perfect. You might also want to add a second parameter of `[]` to useEffect so this podcast is only loaded when this component is mounted.

afraser replied:
I _think_ that'd get rid of the need for the `loaded` state... but practically speaking you'd probably want to keep track of whether it's loading so you can display a spinner or something

afraser commented on src/components/podcast/Podcast.js:
```diff
@@ -0,0 +1,63 @@
+
+
+    return (
+        <div><header><PodcastHeader podcast={podcast} /></header>
```
style nit: I've never seen this style before. Generally every tag gets its own line.

ie:
```jsx
<div>
  <header>
    <PodcastHeader podcast={podcast} />
  </header>
  ...
</div>
```

afraser replied:
Another nit: The `PodcastHeader` should most likely just use a `<header>` tag instead of a `<div>` as it's outermost wrapper.

keegsands replied:
Moved header into the PodcastHeader component. Updated all tags so they get their own line.

afraser commented on src/components/podcast/Podcast.js:
```diff
@@ -0,0 +1,63 @@
+            {episodes && episodes.length > 1 ? <button type="button" onClick={handleOrderFlip} className='flip-button'>Flip Order</button> : null}
+            <ul>
+                {episodes.map((value, index) => {
+                    return <div key={index}><Episode episode={value} /></div>
```
any reason you're wrapping in a div instead of just this?:

```jsx
<Episode key={index} episode={value} />
```

afraser replied:
... then again, that div should actually be a `<li>` since you're in a list here.

afraser replied:
...also, this is another place where I really think implicit returns are cleaner

afraser commented on src/components/podcast/Tile.js:
```diff
@@ -0,0 +1,10 @@
+ * Component for displaying the tile of a podcast
+ * @param {*} podcast podcast object
+ */
+const PodcastTile = ({ podcast }) => {
```
IMO an implicit return reads cleaner for stateless components like this, but YMMV:

```jsx
const PodcastTile = ({ podcast }) =>
    <div>
      <a href={podcast.name}>{podcast.title}</a>
    </div>
```

Another consideration (here it def doesn't matter), is when to avoid object props like podcast here.

A much more reusable component would be one that takes `href` and `title` props.

ie:
```jsx
<PodcastTitle title={podcast.title} href={podcast.name} />
```

or if you keep use `name` instead of `href` as a prop you could do:
```jsx
<PodcastTitle {...podcast} />
```
... I like shortcuts like this but I generally think the above is better because it's much clearer what's getting used, and what it is being used for.

The general guideline I follow is if I'm creating a "container" or "view" component, it's likely that it'll _need_ to accept entire objects as props (eg: when injecting something from a global state store)... everything else (especially components that are likely to be reused) I lean towards named props. All that said, if you're passing more than 5 or 6 props that are all part of the same model, it's time to break this rule. eg: a `PodcastCard` component would almost certainly need quite a bit of information that it wouldn't make sense to split out into props.

afraser commented on src/components/podcast/Header.js:
```diff
@@ -0,0 +1,10 @@
+
+/**
+ * Component for displaying the header of a podcast
+ * @param {*} props podcast object
```
Love seeing js doc tags for components. Very nice. It's also good practice to use PropTypes for all your components to help other people (including future-you) know when they're passing an invalid prop for some reason or another.

afraser commented on src/components/podcast/episode/Episode.js:
```diff
@@ -0,0 +1,12 @@
+ */
+const Episode = ({episode}) => {
+    return (
+        <div><a href={episode.link}><h3>{episode.title}</h3></a> <hr /><div dangerouslySetInnerHTML={{ __html: episode.descriptionText }} /></div>
```
Many teams would force you to do this another way since `dangerouslySetInnerHTML` is a potential security risk. From what I can see in your sample data, there isn't any html in your description text anyway so you should be able to just do `<div>{dangerouslySetInnerHTML}</div>` ... that said you could alternately run this through something like `DOMPurify`

afraser commented on src/components/podcast/Podcast.js:
```diff
@@ -0,0 +1,63 @@
+     * If the feed hasn't been loaded go out and load it.  Set the podcast
+     * state and loaded to true.
+     */
+    async function loadFeed(feedFile) {
```
Generally speaking I don't see people loading data in components under the components/ directory of a project because usually that means those components are meant to be reusable which might mean rendering a bunch of them on a page... I realize that's not the case with this component, but I thought I'd call it out anyway. I think for the purposes of this app, this whole component should just be merged into screens/podcast/Detail.js

keegsands committed:
Code Review Fixes

Several updates based on code review from Adam
Code review Fixes

More improvements based on code review
Adding prop-types for components
Changed some components to be implicit returns.
Code Review Fixes

Merged Podcast into the detail component because it had too much logic for a component.
Code Review Changes

One last implied return value

keegsands committed:
Code Review Fixes

Using DOM Purify to clean up some raw HTML before it is inserted into the page

keegsands committed:
Code Review Fixes

Cleaning up some warnings with the loading & loaded logic

keegsands committed:
Switch to useReducer

Changed podcast detail page to use useReducer instead of useState because it was getting too complex.

dependabot[bot] committed:
Bump dompurify from 1.0.11 to 2.0.14

Bumps [dompurify](https://github.com/cure53/DOMPurify) from 1.0.11 to 2.0.14.
- [Release notes](https://github.com/cure53/DOMPurify/releases)
- [Commits](https://github.com/cure53/DOMPurify/compare/1.0.11...2.0.14)

Signed-off-by: dependabot[bot] <support@github.com>

keegsands committed:
Merge pull request #2 from keegsands/dependabot/npm_and_yarn/dompurify-2.0.14

Bump dompurify from 1.0.11 to 2.0.14
